textdb/textdb-dataflow/src/main/java/edu/uci/ics/textdb/dataflow/keywordmatch/KeywordMatcher.java

```diff
@@ -0,0 +1,168 @@
+import edu.uci.ics.textdb.api.common.IField;
+import edu.uci.ics.textdb.api.common.IPredicate;
+import edu.uci.ics.textdb.api.common.ITuple;
+import edu.uci.ics.textdb.api.common.Schema;
+import edu.uci.ics.textdb.api.dataflow.IOperator;
+import edu.uci.ics.textdb.api.dataflow.ISourceOperator;
+import edu.uci.ics.textdb.common.exception.DataFlowException;
+import edu.uci.ics.textdb.common.field.Span;
+import edu.uci.ics.textdb.common.field.StringField;
+import edu.uci.ics.textdb.common.field.TextField;
+import edu.uci.ics.textdb.common.utils.Utils;
+import edu.uci.ics.textdb.dataflow.common.KeywordPredicate;
+
+/**
+ *  @author prakul
+ *
+ */
+public class KeywordMatcher implements IOperator {
+    private final KeywordPredicate predicate;
+    private ISourceOperator sourceOperator;
+    private String regex;
+    private Pattern pattern;
+    private ArrayList<Pattern> patternList;
+    private Matcher matcher;
+    private List<Span> spanList;
+    private Schema schema;
+    private Schema spanSchema;
+    private int positionIndex; // next position in the field to be checked.
+    private int spanIndexValue; // Starting position of the matched dictionary
+    private String documentValue;
+    private  String fieldName;
+    private  String queryValue;
+    private  List<Attribute> attributeList;
+    private  ArrayList<String> queryValueArray;
+    private ITuple sourceTuple;
+    private List<IField> fieldList;
+    private boolean foundFlag;
+    private boolean schemaDefined;
+
+
+    public KeywordMatcher(IPredicate predicate, ISourceOperator sourceOperator) {
+        this.predicate = (KeywordPredicate)predicate;
+        this.sourceOperator = sourceOperator;
+    }
+
+    @Override
+    public void open() throws DataFlowException {
+        try {
+            sourceOperator.open();
+            queryValue = predicate.getQuery();
+            attributeList = predicate.getAttributeList();
+            queryValueArray = predicate.getTokens();
+            patternList = new ArrayList<Pattern>();
+            for(String token : queryValueArray ){
+                regex = "\\b" + token.toLowerCase() + "\\b";
+                pattern = Pattern.compile(regex);
+                patternList.add(pattern);
+            }
+            positionIndex = 0;
+            foundFlag = false;
+            schemaDefined = false;
+
+            spanList = new ArrayList<>();
+
+        } catch (Exception e) {
+            e.printStackTrace();
+            throw new DataFlowException(e.getMessage(), e);
+        }
+    }
+
+    @Override
+    public ITuple getNextTuple() throws DataFlowException {
+        try {
+            sourceTuple = sourceOperator.getNextTuple();
+            if(sourceTuple == null){
+                return null;
+            }
+            fieldList = sourceTuple.getFields();
+            spanList.clear();
+            if(!schemaDefined){
+                schemaDefined = true;
+                schema = sourceTuple.getSchema();
```

Comment:
This code calculates schema only once (assumes same schema). For creating schema in open() function, we would have to call getNextTuple() in open to obtain a sourceTuple, which is not recommended.